### PR TITLE
Add SWLOR custom TLK editor

### PR DESCRIPTION
**Pull request by zunath**
## Summary

- add a singleton docked TLK Editor for `SWLOR_Haks/sw_tlk/sw_tlk.tlk.json`
- support raw row/custom StrRef navigation, filtering, previous/next find, exact matching, and safe first/next blank allocation
- index custom TLK references across SWLOR 2DAs and protect referenced blank rows
- support multiline row editing plus lossless multi-row copy/paste
- add Open TLK Row shortcuts to supported custom-StrRef LocString fields
- save JSON and generated TLK binary as a verified paired transaction, then publish the exact accepted binary generation to open toolset labels
- add a managed text-only TLK V3.0 writer with shared reader/writer safety limits

## Validation

- `dotnet build SWLOR.Toolset.Tests\SWLOR.Toolset.Tests.csproj -p:RunPostBuildEvent=Never --no-restore`
- `dotnet test SWLOR.Toolset.Tests\SWLOR.Toolset.Tests.csproj --no-build --filter "FullyQualifiedName~Tlk|FullyQualifiedName~ViewLocator" -p:RunPostBuildEvent=Never` — 66 passed
- `dotnet test SWLOR.NWN.Formats.Tests\SWLOR.NWN.Formats.Tests.csproj --no-build -p:RunPostBuildEvent=Never` — 75 passed
- broad Toolset run: 2,963 passed, 1 skipped; four pre-existing unrelated corpus/serializer assertions reproduced in isolation
- repeated independent writer/save and UI/lifecycle reviews completed with no actionable findings

The SWLOR_Haks submodule pointer is unchanged.

## Summary by CodeRabbit

* **New Features**
  * Added a dedicated TLK editor accessible from the Tools menu.
  * Edit, filter, navigate, copy, paste, undo, and redo localized text entries.
  * Add or remove sparse rows and open custom TLK rows from localized-string fields.
  * View repository-wide references and usage details.
  * Save and publish validated TLK updates, with loading progress and cancellation.
  * Refresh localized labels and dropdown options after TLK changes.

* **Bug Fixes**
  * Improved validation for encoding, references, file changes, and allocation limits.

* **Tests**
  * Expanded coverage for editing, persistence, navigation, references, encoding, and boundary conditions.

### CODE REVIEW

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the reference-indexing issue in `17104337f8`: blank allocation now indexes custom StrRefs across repository text sources (including module JSON and C#), fails closed when an eligible reference file cannot be scanned, and includes focused/corpus regression coverage. The editor labels now use **First Blank** and **Next Blank** with no user-facing “safe” wording. Toolset build passes and focused TLK/ViewLocator tests pass (67/67).

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed both findings in `0a81546939`: new rows can no longer skip an earlier Blank through typing or multi-row paste, while a paste may cross the old end only when it fills every earlier Blank; TLK publication now invalidates the shared lookup/resource caches and rebuilds already-open schema dropdowns and locstring labels. Build passes and focused TLK/dropdown/view tests pass (71/71).

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed both findings in `5cd8a77407`: Blank navigation, clear, paste, and save now rebuild the reference index; save rechecks cleared rows that became referenced or lost complete reference coverage; editing an existing row to empty is rejected and directs the builder to **Clear row**, preserving the confirmation path. Build passes and focused TLK/dropdown/view tests pass (74/74).

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the latest review round in `2f848d8ce0`:

- kept the blank-navigation labels concise as **First Blank** and **Next Blank**
- blocked direct typing into known referenced blank rows and rechecked newly populated rows at save time
- made close wait for blank search, clear, paste, reference refresh, and save operations
- handled CR-only trailing clipboard separators without clearing the following row
- logged reference-scan failures with their exception and path
- added focused regressions for all four behaviors

Validation: Toolset test project builds successfully; 78/78 focused TLK/ViewLocator/dropdown tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the latest Codex review in `ef9951bda4`:

- compare the current TLK document against a captured last-saved entry snapshot, so undoing across the save boundary cannot bypass clear/populate confirmation
- serialize blank search, clear, grid paste, reference refresh, and save through one operation gate, so Save All waits for active TLK edits
- added regressions for save → undo → newly referenced clear and for Save racing an in-progress paste

Validation: Toolset test project builds successfully; 80/80 focused TLK/ViewLocator/dropdown tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the latest Codex finding in `66d4428ca5`:

- TLK reads now reserve twice the encoded byte length before decoding, accounting conservatively for UTF-16 managed string storage
- TLK writes use the same decoded UTF-16 budget so the writer cannot generate a file the reader rejects at the shared limit
- added a regression proving oversized single-byte text is rejected before string allocation and tightened the writer boundary test

Validation: 11/11 TLK format tests pass; Toolset test project builds successfully; 80/80 focused TLK/ViewLocator/dropdown tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the concurrent CodeRabbit findings in `691ac5bfa2`:

- reader allocation accounting now uses `Encoding.GetCharCount` over the single validated byte span, then decodes that same span
- added near-budget round-trip coverage for TLK language IDs 128, 129, 130, and 131
- added the missing Serilog warning for malformed-2DA raw-text fallback scan failures

Validation: 15/15 TLK format tests pass; Toolset test project builds successfully; 80/80 focused TLK/ViewLocator/dropdown tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the stale-head Codex findings at current head `470e0d428c`:

- `691ac5bfa2` already aligned reader/writer DBCS budgeting on decoded UTF-16 character count and added all four near-limit language regressions
- moved pending StrRef assignment out of the already-open TLK editor path, explicitly clears stale handled requests, preserves navigation requests while first-open is still loading, and avoids retaining requests when the TLK source is unavailable
- added a lifecycle regression for navigating an already-open editor with a stale pending request

Validation: 15/15 TLK format tests pass; Toolset test project builds successfully; 81/81 focused TLK/ViewLocator/dropdown tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed all three findings from the Codex review of `470e0d428c` in `a4a7fa1244`:

- reference refreshes now retain per-file scan results and reread only added or changed sources; unchanged TLK actions no longer reread the repository corpus
- repository TLK generations published by the editor remain authoritative across later module HAK/TLK assignment reloads
- external TLK reloads keep the active filter applied and select a matching fallback row instead of silently showing an unfiltered range

Added focused regressions for incremental reference refresh, TLK publication precedence, and filtered external reload behavior.

Validation: Toolset test project builds successfully; 84/84 focused TLK/ViewLocator/dropdown tests pass; 15/15 TLK format tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed both findings from the Codex review of `a4a7fa1244` in `607dec1c53`:

- save now revalidates newly appended rows after refreshing references, so a lower blank freed since the editor opened must be used before the append can persist
- production backends establish a verified in-memory generation from the current repository JSON/binary pair, and opening the editor immediately republishes that generation so a previously published older generation cannot remain active
- clarified the runtime precedence documentation for editor-published repository TLKs versus module-selected TLKs

Added regressions for save-time gap revalidation, reopening a newer repository generation, and the complete EditorService open/publication path.

Validation: Toolset test project builds successfully; 87/87 focused TLK/ViewLocator/dropdown tests pass; 15/15 TLK format tests pass.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the remaining TLK publication refresh issue in `60966474d9`:

- cached catalog LocString metadata is re-resolved without rereading ~18k blueprint files
- Module Explorer and Search refresh when catalog labels change
- the active Properties panel rebuilds its materialized TLK-backed values
- standard palette caches are invalidated and open palette/category views are notified
- existing open blueprint/area editors and game-resource lookup caches still refresh
- added a regression covering catalog, Properties, and palette-cache refreshes

Validation: Toolset build succeeded; focused Toolset tests 88/88; TLK format tests 15/15.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed all three findings from the Codex review of `60966474d9` in `ee424cb383`:

- trigger, waypoint, and sound editors now rebuild every materialized choice row after TLK publication; selected specialized editors inside open areas refresh as well
- catalog insertion and TLK-label refresh share a lock and insertion re-resolves the cached LocString source, closing the initial-build publication race
- every decoded sound ResRef now receives a separate managed-string allocation charge

Added deterministic regressions for the catalog race and sound allocation ceiling, and expanded palette-choice invalidation coverage to the TLK refresh path.

Validation: Toolset build succeeded; focused Toolset tests 90/90; rebuilt TLK format tests 16/16.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the asynchronous row-selection finding in `bbcbc81901`:

- Clear Row captures its target ID before refreshing references and uses that fixed ID through confirmation and mutation
- Next Blank now likewise starts from the row on which the command was invoked
- added a deterministic regression that navigates to another TLK row while Clear Row is blocked in reference refresh and verifies only the original row is cleared

Validation: Toolset build succeeded; focused Toolset tests 90/90.

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the latest Codex findings in aeae26e58c:

- TLK allocation accounting now includes managed string overhead and decoded-range dictionary cost for each unique decoded text value, with reader/writer limits kept aligned.
- TLK snapshots now parse bytes and derive fingerprints from the exact same captured file generation, preventing mixed-generation baselines during external writes.

Validation:
- Focused SWLOR.Toolset tests: 90/90 passed
- TLK format tests: 17/17 passed
- Both affected projects build successfully

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Addressed the latest Codex finding in b9cf89fc05:

- TLK publication now rebuilds materialized dynamic choices in open door, creature, item, and merchant editors.
- Creature deferred-choice caches and role rows are refreshed as part of the same operation.
- Selected door editors embedded in open area documents are included.
- Regression coverage now verifies refreshed palette and non-palette choices across all specialized editor types.

Validation:
- Toolset build succeeded
- Focused Toolset tests: 91/91 passed
- TLK format tests: 17/17 passed

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Implemented the latest TLK editor feedback in `13ce86d7f0`:

- expanded the references pane and added Source/File/Row/Label/Column visibility with tooltips
- ordered structured 2DA references ahead of repository-text matches
- excluded `.agents`, `.claude`, and `.codex` metadata/worktrees from reference indexing to remove noisy hits and reduce initial load time
- removed the paste guidance text
- added an immediate TLK Editor loading tab with indeterminate progress and close-to-cancel behavior

Verification:
- Toolset build succeeded
- 94 focused Toolset tests passed
- 17 focused TLK format tests passed

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs [hunk 1]
```diff
@@ -276,11 +303,13 @@ private void PublishSnapshot()
         private CatalogEntry? BuildEntry(ResourceType type, string resRef)
         {
             var path = _workspace.GetResourcePath(type, resRef);
+            var key = IdentityKey(type, resRef);
 
             try
             {
                 var bytes = File.ReadAllBytes(path);
                 var metadata = ExtractMetadata(type, bytes);
+                _nameSources[key] = metadata.NameSource;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs]: Fixed in `b11cbf8dee`. `BuildEntry` now returns the entry and `CatalogNameSource` as one result. The initial build publishes both only when its `TryAdd` wins, while `RefreshEntry` updates the source and entry together under `_tlkLabelLock`. Added a deterministic regression covering an older initial parse completing after a newer refresh.

File: SWLOR.NWN.Formats/Tlk/TlkWriter.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                throw new ArgumentException($"TLK entry ID {id} occurs more than once.", nameof(entries));
+            }
+
+            totalTextLength = checked(totalTextLength + encodedText.Length);
+            totalDecodedTextBytes = checked(
+                totalDecodedTextBytes +
+                TlkFormatLimits.EstimatedManagedStringOverheadBytes +
+                TlkFormatLimits.EstimatedDecodedRangeDictionaryBytes +
+                (long)text.Length * sizeof(char));
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.NWN.Formats/Tlk/TlkWriter.cs:82]: No code change for this refactor suggestion. The standalone formats layer already includes the entry count/allocation estimate/budget in its rejection messages, and the Toolset surfaces the complete rejection reason at its user-facing load/save boundaries without logging decoded text. Adding a new structured exception contract solely to duplicate those values into Serilog is outside this editor change and is not needed for correctness.

**Comment by zunath** [comment]
Addressed the latest functional review findings in `b11cbf8dee`:

- catalog entries and their LocString name sources now publish atomically, preventing an older initial catalog parse from overwriting newer refresh metadata
- TLK row openers now recheck the module mutation lock at invocation time
- added deterministic regressions for both races

Validation:
- Toolset build succeeded
- 116 focused Toolset tests passed

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs]: Fixed in `b11cbf8dee`. The older initial parse no longer writes `_nameSources` from `BuildEntry`; it publishes the entry and source together under `_tlkLabelLock` only when `TryAdd` succeeds. `RefreshEntry` likewise publishes both together. `BlueprintCatalog_RefreshEntryKeepsItsNameSourceWhenInitialBuildFinishesLater` deterministically covers this exact ordering.

**Comment by zunath** [comment]
Added explicit TLK row controls in commit `abfeaf3b4f`:

- **Add row** selects the first unreferenced blank TLK slot (preserving existing StrRefs) and focuses the text editor.
- **Remove row** blanks the selected entry without shifting or renumbering any later row; referenced rows retain the existing confirmation flow.
- Remove is disabled for an already blank row, and both operations remain integrated with undo/redo.

Verification:
- `dotnet build SWLOR.Toolset.Tests\SWLOR.Toolset.Tests.csproj -p:RunPostBuildEvent=Never --no-restore`
- 117 focused toolset tests passed
- 17 TLK format tests passed

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset/Editors/Tlk/TlkEditorDocumentViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,1180 @@
...
+    private async Task AddRowCoreAsync()
+    {
+        if (!await TryRefreshReferencesAsync().ConfigureAwait(true))
+            return;
+        if (!CanAllocateBlank())
+            return;
+        var id = _backend.FindFirstAvailableBlank();
+        SelectId(id, clearFilter: true);
+        NavigationStatus = $"Row {id} is ready for a new TLK entry.";
+        EntryEditRequested?.Invoke();
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset/Editors/Tlk/TlkEditorDocumentViewModel.cs]: Fixed in `ac33549ed6`. Add row now returns its success result through `RunBusyOperationAsync` and raises `EntryEditRequested` only after that wrapper has executed `EndBusyOperation`, so the text editor is enabled before the view requests focus. The regression asserts `IsBusy == false` inside the focus-request event.

File: SWLOR.Toolset.Domain/GameData/Tlk/TlkDocument.cs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            Language = language;
+            _entries = entries;
+        }
+
+        public int Language { get; }
+
+        public int Count => _entries.Count;
+
+        /// <summary>The greatest populated row id, or -1 when the document has no entries.</summary>
+        public int MaxEntryId => _entries.Count == 0 ? -1 : Entries[^1].Id;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset.Domain/GameData/Tlk/TlkDocument.cs]: Fixed in `ac33549ed6`. `TlkDocument` now maintains `_maxEntryId` across sparse adds and removals, making ordinary `MaxEntryId` reads constant-time and independent of the invalidated `Entries` snapshot. Added coverage for lower/higher inserts, non-maximum removal, maximum removal, and the empty document case.

**Comment by zunath** [comment]
Addressed both Codex findings in `ac33549ed6`:

- Add-row focus is requested only after the busy state is released, with a regression checking the event observes `IsBusy == false`.
- `TlkDocument.MaxEntryId` is now maintained independently, avoiding a full 22k-entry snapshot rebuild on each keystroke.
- Both review threads were replied to and resolved.

Validation:
- Toolset build succeeded
- 118 focused Toolset tests passed
- 17 TLK format tests passed

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Simplified the TLK references heading in `5bce7a467f`:

- `References (2DA first)` is now simply `References`.

Validation:
- Toolset build succeeded
- Focused Avalonia render test passed

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
Reordered the TLK row controls in `9dccdddb0b`:

- Toolbar order is now **Next Blank**, **Add row**, **Remove row**.
- Add and Remove are adjacent immediately to the right of Next Blank.
- Added a rendered-view regression covering the button columns.

Validation:
- Toolset build succeeded
- Focused Avalonia toolbar-layout test passed

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs [hunk 5]
```diff
@@ -221,17 +248,33 @@ private void Build(Action<int, int>? onProgress)
...
+                      existingSource == built.NameSource;
+                if (_indexedEntries.TryGetValue(key, out var existing) &&
+                    existing == entry && sourceUnchanged)
+                    return existing;
+
+                if (built.NameSource != null)
+                    _nameSources[key] = built.NameSource;
+                else
+                    _nameSources.TryRemove(key, out _);
+                _indexedEntries[key] = entry;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset.Domain/Workspace/BlueprintCatalog.cs:282]: Fixed in `79d48d2b9d`. `RefreshEntry` now rechecks the deletion tombstone while holding `_tlkLabelLock`, removes any name source, and returns without publishing when a concurrent `RemoveEntry` won. Added a deterministic blocked-resolver race test proving the deleted entry stays absent. Focused catalog and toolbar tests pass 20/20.

**Comment by zunath** [comment]
Addressed the latest review finding in `79d48d2b9d`:

- An in-flight catalog refresh now rechecks the deletion tombstone under the publication lock and cannot restore a deleted resource.
- Added a deterministic refresh-vs-remove regression.
- Replied to and resolved the review thread.

Validation:
- Toolset build succeeded
- 20 focused catalog and toolbar tests passed

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset.Domain/GameData/Tlk/TlkReferenceIndex.cs [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+        private sealed record SourceDescriptor(string Path, string DisplayName, SourceKind Kind);
+
+        private sealed record SourceSnapshot(
+            SourceDescriptor Descriptor,
+            SourceFingerprint Fingerprint,
+            TlkReferenceUsage[] Usages,
+            string? Warning);
+
+        private readonly record struct SourceFingerprint(long Length, DateTime LastWriteUtc)
+        {
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset.Domain/GameData/Tlk/TlkReferenceIndex.cs]: Fixed in `1ea2a6cac8`. Cached reference-source fingerprints now include a SHA-256 content hash, with cancellation checks while hashing. A regression replaces a 2DA with equal-length content, restores its original timestamp, and verifies refresh discards the old StrRef and indexes the new one. Focused reference-cache, catalog, and toolbar tests pass 25/25.

**Comment by zunath** [comment]
Addressed the latest Codex finding in `1ea2a6cac8`:

- Reference-source cache fingerprints now include SHA-256 content hashes.
- Same-length replacements with preserved timestamps can no longer reuse stale StrRef scans.
- Added exact regression coverage and resolved the review thread.

Validation:
- Toolset build succeeded
- 25 focused reference-cache, catalog, and toolbar tests passed

**Comment by zunath** [comment]
@codex review

**Comment by zunath** [comment]
@codex review

File: SWLOR.Toolset/Editors/Tlk/TlkEditorBackend.cs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    public int FindFirstAvailableBlank() => _document.FindFirstAvailableBlank(_references);
+    public int FindNextAvailableBlank(int currentId) =>
+        _document.FindNextAvailableBlank(currentId, _references);
+    public void RefreshReferences()
+    {
+        _references = _references.Refresh(_source.TwoDaDirectory, _source.RepositoryRoot);
+    }
+
+    public bool HasExternalChange() =>
+        !_jsonFingerprint.Matches(JsonPath) || !_binaryFingerprint.Matches(BinaryPath);
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Toolset/Editors/Tlk/TlkEditorBackend.cs]: Fixed in `5658bf2b8`. The final external-change guard now captures both fingerprints forward and revalidates them in reverse order, failing closed if either file changes during the complete pair check. Post-commit baselines retain the exact bytes written. A deterministic regression replaces the JSON after its first capture while the binary is being checked; the unstable pair is rejected. Focused TLK editor tests pass 39/39.

**Comment by zunath** [comment]
@codex review
